### ISSUE TITLE
splice not removing data from Polymer Collections?

### ISSUE THREAD

**Issue by akc42** [Polymer/polymer#3865]
### Description
Polymer throws TypeError: Cannot read property 'length' of undefined in polymer.html:2315

This is a setting up a property that drives a set of nested don-repeats representing all the days of a month (and a bit). in an appointment calendar.  When the user changes month, we need to clear the array down and then build it back up again for the new month.

In debugging the problem, it occurs the second pass through when using path 'myarray.#0'
at approximately (ie on my version of polymer which is probably older than master) line 2182 of polymer.html there is a line which is attempting to fetch the part

`prop = Polymer.Collection.get(array).getItem(part);`

in the getItem(part) it is looking in the collections `store` with key = 0, but the actual entry is stored at key 6 and keys 0-5 don't exists.   I suspect `this.splice` call made to reset the array has removed some entries but not adjusted the index.  I am afraid my knowledge of the Polymer code doesn't allow me to go much further.

#### Live Demo

https://jsbin.com/boleko/edit?html,console,output

#### Steps to Reproduce

1.Create an array property and built it up using this.push and paths, so that changes are notified
2. Use this.spilce('myarray',0) to clear the array out
3 Rebuild array (normally with new data, but current jsbin uses the same)

#### Expected Results

The setup completes successfully

#### Actual Results

An error is thrown TypeError: Cannot read property 'length' of undefined in polymer.html:2315

### Browsers Affected

- [ X] Chrome
- [ ] Firefox
- [ ] Edge
- [ ] Safari 9
- [ ] Safari 8
- [ ] IE 11

### Versions

- Polymer: v1.6.1
- webcomponents: v0.7.22

**Comment by kaste** [comment]
There is an issue in your JSBin. When you do 
```
path = 'myarray.#' + i;
for(var j = 0; j < 7; j++) {
  this.push(path,{id:j});
};
```
you use the `#` key notation.  But keys are kept to be secret. They are stable even if you wipe the complete array (via `splice` e.g.). In your example you're probably good to go, if you just leave the `#` out. That is your path becomes just `myarray.0` etc. 

Otherwise if it gets hairy, you might need to ask `Polymer.Collections` to return you the correct key for a given item. From the top of my head this is something like `Polymer.Collections.get(this.myarray).getKey(<...item...>)`.

**Comment by akc42** [comment]
Thanks - taking the '#' out really solved it.  I spent hours yesterday trying to get to the bottom of how to manipulate the arrays so that my dom-repeat templates would notice the change.  I found it very difficult to get to the bottom of what exactly the path notation should be.

Taking the #'s out has made it work beautifully - even when an ajax call later returns lots more additional data and it just updates the screen seamlessly.  

Thanks
